frame/bridge/ethereum/issuing/src/lib.rs

```diff
@@ -21,275 +21,347 @@
+			proof: EthereumReceiptProofThing<T>,
+		) -> DispatchResultWithPostInfo {
 			log::debug!("start to redeem erc20 token");
-			let user = ensure_signed(origin)?;
-			let (tx_index, ethlog) = Self::verify_and_parse_proof(
-				Abi::backing_event(),
-				proof)?;
-			let backing_address = EthereumBackingAddress::get();
-			let input = Self::abi_encode_token_redeem(ethlog)?;
+			ensure_signed(origin)?;
+			let tx_index = T::EthereumRelay::gen_receipt_index(&proof);
+			ensure!(
+				!VerifiedIssuingProof::<T>::contains_key(tx_index),
+				<Error<T>>::AssetAR
+			);
+			let verified_receipt = T::EthereumRelay::verify_receipt(&proof)?;
+			let backing_address = EthereumBackingAddress::<T>::get();
+			let lock_info =
+				EthereumBacking::parse_locking_event(&verified_receipt, &backing_address)
+					.map_err(|_| Error::<T>::DecodeEventFailed)?;
+			let input = mtf::encode_cross_receive(
+				lock_info.mapped_address,
+				lock_info.recipient,
+				lock_info.amount,
+			)
+			.map_err(|_| Error::<T>::InvalidEncodeERC20)?;
 			Self::transact_mapping_factory(input)?;
-			VerifiedIssuingProof::insert(tx_index, true);
-			Self::deposit_event(RawEvent::RedeemErc20(user, backing_address, tx_index));
+			VerifiedIssuingProof::<T>::insert(tx_index, true);
+			Self::deposit_event(Event::RedeemErc20(backing_address, tx_index));
+			Ok(().into())
 		}
-	}
-}
 
-impl<T: Config> Module<T> {
-	/// The account ID of the issuing pot.
-	pub fn dvm_account_id() -> H160 {
-		let account32: AccountId32 = T::PalletId::get().into_account();
-		let account20: &[u8] = &account32.as_ref();
-		H160::from_slice(&account20[..20])
-	}
+		#[pallet::weight(<T as Config>::WeightInfo::mapping_factory_event_handle())]
+		pub fn mapping_factory_event_handle(
+			origin: OriginFor<T>,
+			input: Vec<u8>,
+		) -> DispatchResultWithPostInfo {
+			let caller = ensure_signed(origin)?;
+			ensure!(input.len() >= 8, <Error<T>>::InvalidInput);
+			let factory = MappingFactoryAddress::<T>::get();
+			let factory_id = <T as darwinia_evm::Config>::AddressMapping::into_account_id(factory);
+			ensure!(factory_id == caller, <Error<T>>::AssetAR);
+			let burn_action = &sha3::Keccak256::digest(&BURN_ACTION)[0..4];
+			let register_action = &sha3::Keccak256::digest(&REGISTER_ACTION)[0..4];
+			if &input[4..8] == burn_action {
+				let burn_info =
+					TokenBurnInfo::decode(&input[8..]).map_err(|_| Error::<T>::InvalidInputData)?;
+				ensure!(burn_info.recipient.len() == 20, <Error<T>>::AssetAR);
+				Self::deposit_burn_token_event(
+					burn_info.backing,
+					burn_info.sender,
+					burn_info.source,
+					EthereumAddress::from_slice(burn_info.recipient.as_slice()),
+					burn_info.amount,
+				)?;
+			} else if &input[4..8] == register_action {
+				let register_info = TokenRegisterInfo::decode(&input[8..])
+					.map_err(|_| Error::<T>::InvalidInputData)?;
+				Self::finish_token_registered(register_info.0, register_info.1, register_info.2);
+			} else {
+				log::trace!("Unsupport action!");
```

Comment:
Return an error instead of log a message?

Reply:
We can't return an error here. If so, the contract will be reverted.